### PR TITLE
CSHARP-1501: Drivers must raise an error if response messageLength > ismaster.maxMessageSizeBytes

### PR DESCRIPTION
**Pull request by ksadoff**
https://jira.mongodb.org/browse/CSHARP-1501
~~https://evergreen.mongodb.com/version/5d0d464232f41749162d93a7~~
~~https://evergreen.mongodb.com/version/5d10da1e9ccd4e148cd09c92~~

Ready for interim review.

### CODE REVIEW

**Review by vincentkam — COMMENTED** [review]
Everything looks good to me: just waiting to see what @rstam thinks about the type of exception that should be thrown :)

File: tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs [hunk 1]
```diff
@@ -276,6 +276,45 @@ public void Dispose_should_raise_the_correct_events()
             act.ShouldThrow<InvalidOperationException>();
         }
 
+        [Theory]
+        [ParameterAttributeData]
+        public void ReceiveMessage_should_throw_error_when_message_is_an_invalid_size(
```
Comment (vincentkam) [inline tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs]: This is totally optional, but a slightly more modern take might be to declare the test as:
`public async void` and use the `await` keyword for the `async` variant. See [this test](https://github.com/mongodb/mongo-csharp-driver/blob/7d426a8243532983aee819d2b227c873d80df75d/tests/MongoDB.Driver.Tests/PinnedShardRouterTests.cs#L55) for an example. Completely optional, as the way you have it here definitely works, so we could save it for next time 😄

File: tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs [hunk 2]
```diff
@@ -276,6 +276,45 @@ public void Dispose_should_raise_the_correct_events()
...
+            bool async)
+        {
+            using (var stream = new BlockingMemoryStream())
+            {
+                var bytes = BitConverter.GetBytes(length);
+                stream.Write(bytes, 0, bytes.Length);
+                stream.Seek(0, SeekOrigin.Begin);
+                var encoderSelector = new ReplyMessageEncoderSelector<BsonDocument>(BsonDocumentSerializer.Instance);
+
+                Action received;
```
Comment (vincentkam) [inline tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs]: Hm, food for thought: as this is an `Action`, would a verb work better as a name than an adjective? Perhaps something like `ReceiveMessage`?

File: tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs (hunk 1 above)
Comment (rstam) [reply tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs]: Either way there is a mismatch because we are using a single test to test both the sync and async variants of a method.

Either:

1. we are using GetAwaiter().GetResult() when testing the async method
2. we are blocking in an async test method when testing the sync method (blocking is bad in async code)

In production code both those scenarios would not be tolerated. In test code I don't see much advantage to using XUnit's new support for async test methods unless the test itself is 100% async.

If you can think of reasons why the new async test method is better, elaborate (other than being newer!).

File: tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs (hunk 2 above)
Comment (rstam) [reply tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs]: I assume this was copy/pasted from an existing test method being adapted for the new test.

In brand new tests we would FluentAssertions Record.Exception instead.

File: src/MongoDB.Driver.Core/Core/Connections/BinaryConnection.cs [hunk 3]
```diff
@@ -314,13 +314,22 @@ private async Task OpenHelperAsync(CancellationToken cancellationToken)
             }
         }
 
+        private void EnsureMessageSizeIsValid(Int32 messageSize)
+        {
+            if (messageSize < 0 || messageSize > _description.MaxMessageSize)
+            {
+                // What kind of exception to use here?
+                throw new MongoInternalException("The size of the message is invalid.");
```
Comment (ksadoff) [inline src/MongoDB.Driver.Core/Core/Connections/BinaryConnection.cs]: @rstam Do you have any thoughts about what type of error this function should throw?

File: tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs (hunk 1 above)
Comment (vincentkam) [reply tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs]: Hmm, you make a good point: I agree that both scenarios would not be tolerated in production code, so in that sense they're both "equally unacceptable."

Initially, my thought process was that they're both equally unacceptable, then I'd prefer to use the version that's slightly newer and results in slightly more concise code. 

As a counterpoint to my own preference for `async void`, before C#7 introduced `async main`, I think standard practice was to call use define an async helper and to call the helper with `GetAwaiter.GetResult()`, so there's at least an accepted history behind using `GetAwaiter.GetResult()` in a synchronous method.

Just as a discussion point, as an alternative, I suppose we could have a sync and async version of each test defined, with a helper method returning Assert, Arrange, Act Actions, but I'm not sure we'd gain much over the current implementation.

TLDR: No action needed, resolving this thread :)

File: tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs (hunk 2 above)
Comment (vincentkam) [reply tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs]: Ah true, I had missed this :)

File: tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs [hunk 4]
```diff
@@ -276,6 +276,45 @@ public void Dispose_should_raise_the_correct_events()
...
+                else
+                {
+                    _mockStreamFactory.Setup(f => f.CreateStream(_endPoint, CancellationToken.None))
+                        .Returns(stream);
+                    _subject.Open(CancellationToken.None);
+
+                    invalidSizeException = Record.Exception( () =>  _subject.ReceiveMessage(10, encoderSelector, _messageEncoderSettings, CancellationToken.None));
+                }
+
+                Assert.NotNull(invalidSizeException);
```
Comment (vincentkam) [inline tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs]: Sorry for the style confusion Katie! I believe that this line and the line below it should probably use Fluent Assertions ([example](https://github.com/mongodb/mongo-csharp-driver/blob/48e7a08bb4069cdbcdfb4f12459db7001a842abd/tests/MongoDB.Driver.Core.Tests/Core/Operations/EndTransactionOperationTests.cs#L43))

**Review by vincentkam — CHANGES_REQUESTED** [review]

**Comment by vincentkam** [comment]
I think we're probably ready for team review!

**Comment by ksadoff** [comment]
Ready for review

**Comment by rstam** [comment]
Starting to review.

Remember when you squash and rebase that commit messages need to start with the JIRA ticket number. The text is usually just the one line summary from the JIRA ticket itself.

For example:

    CSHARP-1501: Drivers must raise an error if response messageLength > ismaster.maxMessageSizeBytes.

File: src/MongoDB.Driver.Core/Core/Connections/BinaryConnection.cs (hunk 3 above)
Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Connections/BinaryConnection.cs]: When we are parsing or decoding the usual exception to throw when the data format is wrong is a `FormatException`. Let's throw that here too.

File: src/MongoDB.Driver.Core/Core/Connections/BinaryConnection.cs [hunk 5]
```diff
@@ -314,13 +314,22 @@ private async Task OpenHelperAsync(CancellationToken cancellationToken)
...
+            }
+        }
```
Comment (rstam) [inline src/MongoDB.Driver.Core/Core/Connections/BinaryConnection.cs]: Blank line between methods.

Put method in alphabetical order.

File: tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs [hunk 6]
```diff
@@ -276,6 +276,46 @@ public void Dispose_should_raise_the_correct_events()
             act.ShouldThrow<InvalidOperationException>();
         }
 
+        [Theory]
+        [ParameterAttributeData]
+        public void ReceiveMessage_should_throw_error_when_message_is_an_invalid_size(
+            [Values(-1, 16000005)]
```
Comment (rstam) [inline tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs]: I was expecting 48000001 as the second test value (48000000 + 1).

In order to use that value let's change the `BinaryConnectionTests` constructor to use some more realistic setups.

Lines 54-67 of the constructor could be refactored to have more realistic values for isMasterResult and buildInfoResult:

    _endPoint = new DnsEndPoint("localhost", 27017);
    var serverId = new ServerId(new ClusterId(), _endPoint);
    var connectionId = new ConnectionId(serverId);
    var isMasterResult = new IsMasterResult(new BsonDocument { { "ok", 1 }, { "maxMessageSizeBytes", 48000000 } });
    var buildInfoResult = new BuildInfoResult(new BsonDocument { { "ok", 1 }, { "version", "2.6.3" } });
    var connectionDescription = new ConnectionDescription(connectionId, isMasterResult, buildInfoResult);

    _mockConnectionInitializer = new Mock<IConnectionInitializer>();
    _mockConnectionInitializer
        .Setup(i => i.InitializeConnection(It.IsAny<IConnection>(), CancellationToken.None))
        .Returns(connectionDescription);
    _mockConnectionInitializer
        .Setup(i => i.InitializeConnectionAsync(It.IsAny<IConnection>(), CancellationToken.None))
        .Returns(Task.FromResult(connectionDescription));

File: tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs [hunk 7]
```diff
@@ -276,6 +276,46 @@ public void Dispose_should_raise_the_correct_events()
...
+            bool async)
+        {
+            using (var stream = new BlockingMemoryStream())
+            {
+                var bytes = BitConverter.GetBytes(length);
+                stream.Write(bytes, 0, bytes.Length);
+                stream.Seek(0, SeekOrigin.Begin);
+                var encoderSelector = new ReplyMessageEncoderSelector<BsonDocument>(BsonDocumentSerializer.Instance);
+
+                Exception invalidSizeException;
```
Comment (rstam) [inline tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs]: I think it would be acceptable to use just `exception` for this variable.

File: tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs [hunk 8]
```diff
@@ -276,6 +276,46 @@ public void Dispose_should_raise_the_correct_events()
...
+                else
+                {
+                    _mockStreamFactory.Setup(f => f.CreateStream(_endPoint, CancellationToken.None))
+                        .Returns(stream);
+                    _subject.Open(CancellationToken.None);
+
+                    invalidSizeException = Record.Exception( () =>  _subject.ReceiveMessage(10, encoderSelector, _messageEncoderSettings, CancellationToken.None));
+                }
+
+                invalidSizeException.Should().NotBe(null);
```
Comment (rstam) [inline tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs]: The assertion could be a little bit stronger than not null:

        exception.Should().BeOfType<MongoConnectionException>();

**Review by rstam — CHANGES_REQUESTED** [review]

File: tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs [hunk 9]
```diff
@@ -212,6 +212,46 @@ public void Dispose_should_raise_the_correct_events()
             _capturedEvents.Any().Should().BeFalse();
         }
 
+        [Theory]
```
Comment (ksadoff) [inline tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs:215]: Renamed this test and moved it to be alphabetical.

**Comment by ksadoff** [comment]
Ready for re-review.

File: src/MongoDB.Driver.Core/Core/Connections/BinaryConnection.cs [hunk 10]
```diff
@@ -220,6 +220,14 @@ private void Dispose(bool disposing)
             }
         }
 
+        private void EnsureMessageSizeIsValid(Int32 messageSize)
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver.Core/Core/Connections/BinaryConnection.cs]: probably it can be just `int`

File: src/MongoDB.Driver.Core/Core/Connections/BinaryConnection.cs [hunk 11]
```diff
@@ -220,6 +220,14 @@ private void Dispose(bool disposing)
...
+        {
+            if (messageSize < 0 || messageSize > _description.MaxMessageSize)
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver.Core/Core/Connections/BinaryConnection.cs]: what about this requirement:
>If this is the first operation run on the server (e.g. the first ismaster command sent on first connection) use a hard coded value of 48MB, switching to ismaster.maxMessageSizeBytes once you have received a valid ismaster response.

?

File: tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs [hunk 12]
```diff
@@ -212,6 +212,46 @@ public void Dispose_should_raise_the_correct_events()
...
+            {
+                var bytes = BitConverter.GetBytes(length);
+                stream.Write(bytes, 0, bytes.Length);
+                stream.Seek(0, SeekOrigin.Begin);
+                var encoderSelector = new ReplyMessageEncoderSelector<BsonDocument>(BsonDocumentSerializer.Instance);
+
+                Exception exception;
+                if (async)
+                {
+                    _mockStreamFactory.Setup(f => f.CreateStreamAsync(_endPoint, CancellationToken.None))
```
Comment (DmitryLukyanov) [inline tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs]: It can be written in this way which probably looks better:

                    _mockStreamFactory
                        .Setup(f => f.CreateStreamAsync(_endPoint, CancellationToken.None))
                        .ReturnsAsync(stream);
                    _subject.OpenAsync(CancellationToken.None).GetAwaiter().GetResult();
                    exception = Record
                        .ExceptionAsync(async () => await _subject.ReceiveMessageAsync(10, encoderSelector, _messageEncoderSettings, CancellationToken.None))
                        .Result;

File: tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs [hunk 13]
```diff
@@ -212,6 +212,46 @@ public void Dispose_should_raise_the_correct_events()
...
+                {
+                    _mockStreamFactory.Setup(f => f.CreateStream(_endPoint, CancellationToken.None))
+                        .Returns(stream);
+                    _subject.Open(CancellationToken.None);
+
+                    exception = Record.Exception( () =>  _subject.ReceiveMessage(10, encoderSelector, _messageEncoderSettings, CancellationToken.None));
+                }
+
+                exception.Should().BeOfType<MongoConnectionException>();
+                exception.InnerException.Should().BeOfType<FormatException>();
```
Comment (DmitryLukyanov) [inline tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs]: One more possible suggestion:
                
                exception.Should().BeOfType<MongoConnectionException>();
                var e = exception.InnerException.Should().BeOfType<FormatException>().Subject;
                e.Message.Should().Be("The size of the message is invalid.");

File: tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs [hunk 14]
```diff
@@ -53,18 +53,18 @@ public BinaryConnectionTests()
...
-                .Returns(() => Task.FromResult(new ConnectionDescription(
-                    new ConnectionId(serverId),
-                    new IsMasterResult(new BsonDocument()),
-                    new BuildInfoResult(new BsonDocument("version", "2.6.3")))));
+            _mockConnectionInitializer
+                .Setup(i => i.InitializeConnection(It.IsAny<IConnection>(), CancellationToken.None))
+                .Returns(connectionDescription);
+            _mockConnectionInitializer
+                .Setup(i => i.InitializeConnectionAsync(It.IsAny<IConnection>(), CancellationToken.None))
+                .Returns(Task.FromResult(connectionDescription));
```
Comment (DmitryLukyanov) [inline tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs]: `ReturnsAsync`?

**Review by DmitryLukyanov — CHANGES_REQUESTED** [review]

File: tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs (hunk 14 above)
Comment (ksadoff) [reply tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs]: This was a change that @rstam suggested, but it does look like it should be async here.

File: src/MongoDB.Driver.Core/Core/Connections/BinaryConnection.cs (hunk 11 above)
Comment (ksadoff) [reply src/MongoDB.Driver.Core/Core/Connections/BinaryConnection.cs]: If I'm understanding the code correctly, that's dealt with here: https://github.com/mongodb/mongo-csharp-driver/blob/fad2887cd28950ebb50d1039e336026d1f169339/src/MongoDB.Driver.Core/Core/Connections/IsMasterResult.cs#L215

Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Connections/BinaryConnection.cs]: >If I'm understanding the code correctly, that's dealt with here

yes, but this value can be used only when we had at least one `isMaster` response.

When we call `isMaster` first time, `_description` will be `null`, since we haven't had `isMaster` response yet and as a result, we haven't filled `_description` yet. So, since `description` is `null` we will have a null reference exception in this line. 
See test results: https://evergreen.mongodb.com/version/5d10da1e9ccd4e148cd09c92 :) 

Also, as I see, the quoted text says that if you haven't had `isMaster` response yet(`e.g. the first ismaster command sent on first connection`), you need to use hardcoded value in MaxMessageSize which is 48 MB, but as soon as the `ismaster` response is received, you will need to use the value from `isMaster` response (`once you have received a valid ismaster response.`)

**Review by DmitryLukyanov — CHANGES_REQUESTED** [review]

File: src/MongoDB.Driver.Core/Core/Connections/BinaryConnection.cs (hunk 10 above)
Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Connections/BinaryConnection.cs]: I agree. We use Int32 in type names (e.g. Int32Serializer) but just int when declaring variables.

File: src/MongoDB.Driver.Core/Core/Connections/BinaryConnection.cs (hunk 11 above)
Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Connections/BinaryConnection.cs]: > you need to use hardcoded value in MaxMessageSize which is 48 MB

I think the spec is wrong. When we haven't received an isMaster result yet we need to use a default value of 16000000 to be compatible with older servers.

In any case it's probably moot... since I can't imagine an isMaster command or response reaching even the smaller max message size.

But the NullReferenceException needs to be fixed... :)

File: tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs (hunk 14 above)
Comment (rstam) [reply tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs]: Not sure when Moq added support for `ReturnAsync` but looks to me like the following two statements are equivalent:

    Returns(Task.FromResult(connectionDescription))

    ReturnsAsync(connectionDescription)

I'm totally fine with changing to `ReturnAsync`. Thanks for finding this method I did not know about!

File: tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs (hunk 12 above)
Comment (rstam) [reply tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs]: `ExceptionAsync` is another method I did not know about. 

I'm not sure about this one though... 

I don't think we can use `.Result`. We would have to use `.GetAwaiter().GetResult()` to unwrap exceptions properly.

Comment (rstam) [reply tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs]: If we chose to use Xunit's support for `async Task` test methods we could probably use this:

        exception = await Record.ExceptionAsync(() => _subject.ReceiveMessageAsync(10, encoderSelector, _messageEncoderSettings, CancellationToken.None));

**Review by rstam — CHANGES_REQUESTED** [review]

File: src/MongoDB.Driver.Core/Core/Connections/BinaryConnection.cs (hunk 11 above)
Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Connections/BinaryConnection.cs]: Dmitry argues that we should just follow the spec and use 48000000 (and not 16000000).

While it is totally true that the limit used to be 16000000, looking at server JIRA tickets it has been 48000000 since at least 2012-12-26 when the `maxMessageSizeBytes` field was first added to the `isMaster` result.

So it seems safe to ignore the fact that the `MaxMessageSize` once used to be smaller.

You can create a local variable here to work around the `NullReferenceException`:

     var maxMessageSize = _description?.MaxMessageSize ?? 48000000;

**Review by rstam — CHANGES_REQUESTED** [review]

File: src/MongoDB.Driver.Core/Core/Connections/BinaryConnection.cs (hunk 11 above)
Comment (ksadoff) [reply src/MongoDB.Driver.Core/Core/Connections/BinaryConnection.cs]: > I think the spec is wrong. When we haven't received an isMaster result yet we need to use a default value of 16000000 to be compatible with older servers.

Just to clarify, when _description is null, I should be using a hard-coded value of 16MB as maxMessageSize?

Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Connections/BinaryConnection.cs]: Go ahead and just follow the spec and use 48000000.

No servers with the smaller 16000000 MasMessageSize should still exist in the wild.

And as Dmitry argues it's easier to just follow the spec than to try and change the spec.

Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Connections/BinaryConnection.cs]: p.s. 48MB and 48000000 are not the same value.

I tend to interpret 48MB to mean 48 * 1024 * 1204 = 50331648.

File: tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs (hunk 12 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs]: One more new feature for me :) 
I didn't know that `async Task` tests are supported. So probably we need to write tests which include `async` methods in a similar way to this one:
       
        [Theory]
        [ParameterAttributeData]
        public async Task ReceiveMessage_should_throw_a_FormatException_when_message_is_an_invalid_size(
            [Values(-1, 48000001)] int length,
            [Values(false, true)] bool async)

It allows us just to use `await` rather than `GetAvaiter` and so on. Probably we need to even add it into our a google doc with code format questions? Thoughts?

Comment (rstam) [reply tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs]: This was discussed here:

https://github.com/ksadoff/mongo-csharp-driver/pull/6#discussion_r296410104

I certainly wouldn't go changing any existing tests.

I agree that using `async Task` test methods looks attractive, but I'm not 100% sure whether there are any negative side effects from blocking (when calling sync methods) in an async method.

Comment (rstam) [reply tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs]: If you decide to change this new test to be an async test method be sure to use `async Task` and not `async void`.

@vincentkam I did not catch this when you first introduced the two existing tests that use `async void`.

Comment (rstam) [reply tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs]: I'm finding hits on Google that blocking in an async Xunit test method can lead to deadlocks.

Perhaps we should stick to what we've been doing all along, and use sync test methods combined with `GetAwaiter().GetResult()` when calling the async method under test.

Comment (rstam) [reply tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs]: I'm also finding comments that using async Xunit test methods can lead to an unbounded number of tests running in parallel.

Apparently every time a test method hits an await statement it is possible for Xunit to schedule a new test method, even though the current test has not finished.

From the limited understanding I have this can't happen with sync test methods since there is a fixed size thread pool to run sync tests on, so the max number of tests running at the same time is limited by the size of the underlying thread pool.

Nice as this new async Xunit test method feature is I'm getting nervous about using it. Might be better to stick to what's been working.

Comment (vincentkam) [reply tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs]: > If you decide to change this new test to be an async test method be sure to use `async Task` and not `async void`.
> 
> @vincentkam I did not catch this when you first introduced the two existing tests that use `async void`.

Ah you're right, `async Task` would be preferable I think. Fwiw, I vaguely recall that XUnit has special machinery for handling `async void` to [call and capture exceptions](https://stackoverflow.com/questions/23824660/xunit-async-test-not-working-properly).

Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs]: >Might be better to stick to what's been working.

based on possible problems that @rstam mentioned above(and I didn't google any proofs that using `async` and `sync` together is safe), I agree, that probably it's better to use `sync` test methods.

Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs]: >I don't think we can use .Result. We would have to use .GetAwaiter().GetResult() to unwrap exceptions properly.

Agree about it.

File: tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs (hunk 6 above)
Comment (vincentkam) [reply tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs]: Should this be `(48*1024*1024*) + 1` (per the discussion about 48MB vs 48000)?

Comment (rstam) [reply tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs]: No. Oddly enough the MaxMessageSize is 48,000,000, not 48MB.

**Comment by ksadoff** [comment]
Ready for re-review.
Updated Evergreen patch: https://evergreen.mongodb.com/version/5d1501af0ae6060cbfdda9e8

**Review by DmitryLukyanov — APPROVED** [review]
Waiting test results and then LGTM

File: tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs [hunk 15]
```diff
@@ -212,6 +212,48 @@ public void Dispose_should_raise_the_correct_events()
...
+
+                Exception exception;
+                if (async)
+                {
+                    _mockStreamFactory
+                        .Setup(f => f.CreateStreamAsync(_endPoint, CancellationToken.None))
+                        .ReturnsAsync(stream);
+                    _subject.OpenAsync(CancellationToken.None).GetAwaiter().GetResult();
+                    exception = Record
+                        .ExceptionAsync(async () => await _subject.ReceiveMessageAsync(10, encoderSelector, _messageEncoderSettings, CancellationToken.None))
```
Comment (rstam) [inline tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs]: The `async` and `await` keywords seem unnecessary here.

File: tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs [hunk 16]
```diff
@@ -212,6 +212,48 @@ public void Dispose_should_raise_the_correct_events()
...
+                        .GetAwaiter()
+                        .GetResult();
```
Comment (rstam) [inline tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs]: If we're going to use `.GetAwaiter().GetResult()` anyway I don't see any advantage to using `ExceptionAsync` (a search reveals that we *have* used `ExceptionAsync` before but in very few places).

Lines 237-240 could just be:

    exception = Record.Exception(() => _subject.ReceiveMessageAsync(10, encoderSelector, _messageEncoderSettings, CancellationToken.None).GetAwaiter().GetResult());

File: tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs (hunk 9 above)
Comment (rstam) [reply tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs:215]: Ack

**Review by rstam — APPROVED** [review]
I'm willing to LGTM as is because everything should be 100% working.

But I would like all team members to read the comments I just added to the new test method regarding async because we should probably resolve how we think about these methods in order to have a consistent approach going forward.

File: tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs (hunk 15 above)
Comment (vincentkam) [reply tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs]: Doesn't `async ()` create an anonymous `async` function, which then enables the use of the `await` operator? So as written, both `async`/`await` should be needed.

File: tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs (hunk 16 above)
Comment (vincentkam) [reply tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs]: Since we are choosing not to make the test method `async` and are thus unable to `await Record.ExceptionAsync`, then I agree that the above refactoring is perfectly reasonable.

Playing devil's advocate though: Is there any **disadvantage** to using the current format though? Your previous comment mentioned finding hits on Google regarding issues if we block during an `async` test method, which shouldn't be an issue here as the method is declared as a synchronous test method.  IMHO, one minor advantage of the `async () => await...` format is that it makes it exceedingly clear that the method we're testing is an `async` method (and it's "front loaded" onto the statement as opposed to `.GetAwaiter().GetResult()` which happens later).

**Review by vincentkam — APPROVED** [review]
LGTM! 😺

File: tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs (hunk 15 above)
Comment (vincentkam) [reply tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs]: Oops, I think if we're going with this style for this ticket, then we'll need `ConfigureAwait(false)` as part of the `await`
```
                     exception = Record
                        .ExceptionAsync(async () => await _subject.ReceiveMessageAsync(10, encoderSelector, _messageEncoderSettings, CancellationToken.None).ConfigureAwait(false))
                        .GetAwaiter()
                        .GetResult();
```

**Review by vincentkam — CHANGES_REQUESTED** [review]

File: tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs (hunk 15 above)
Comment (rstam) [reply tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs]: Let me rephrase in more detail:

`await` is not necessary because no further work is done after the async call

`async` is not necessary once we remove the `await`

Comment (rstam) [reply tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs]: `ConfigureAwait(false)` is not necessary in the test projects.

It is of course mandatory in the non-test projects.

File: tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs (hunk 16 above)
Comment (rstam) [reply tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs]: Playing devil's advocate though: is there any **advantage** to using `ExceptionAsync`? I would say no, because the overall test method is not an async method, so `ExceptionAsync` is kind of out of place.

Using `ExceptionAsync` also does NOT avoid the need for using `GetAwaiter().GetResult()`. It just forces you to move the `GetAwaiter().GetResult()` to *outside* of the `ExceptionAsync`.

I'd rather call `GetAwaiter().GetResult()` on the original async call than on the `Task<Exception>` task returned by `ExecuteAsync`.

I agree with the comment that it's nice to use `await`, which is why we were tempted to start using async test methods. But in this case, the await is both unnecessary (because no work is done after the await) and at the wrong level (it's nested inside a lambda and not part of the actual test method).

**Review by rstam — COMMENTED** [review]
Responding to Vincent's comments.

File: tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs (hunk 15 above)
Comment (vincentkam) [reply tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs]: Oh fair, we don't need it in the test projects (although I think it doesn't hurt?). Sorry @ksadoff, your code as fine as is!

Re: `async/await`: Oh very cool, you are indeed correct! It's required to have `async` if we want to use `await` but without `await`, of course we don't need `async`! 

(Side comment for gory in depth details in the interest of science/love of C# :) ):
I tried dropping the `await` call and got the following warnings, which of course we count as errors:
```
  BinaryConnectionTests.cs(238, 50): [CS1998] This async method lacks 'await' operators and will run synchronously. Consider using the 'await' operator to await non-blocking API calls, or 'await Task.Run(...)' to do CPU-bound work on a background thread.
  BinaryConnectionTests.cs(238, 54): [CS4014] Because this call is not awaited, execution of the current method continues before the call is completed. Consider applying the 'await' operator to the result of the call.
```

Running some extra tests for fun also confirmed that using `async/await` inside `Record.ExceptionAsync` and not using `async/await` inside `Record.ExceptionAsync` are equivalent :): 
```
 public void AsyncTest()
        {
            Task fail()
            {
                var tcs = new TaskCompletionSource<int>();
                tcs.SetException(new NotImplementedException());
                return tcs.Task;
            }
            var exception = Record
                .ExceptionAsync( async () =>  await fail())
                .GetAwaiter()
                .GetResult();
            exception.Should().BeOfType<NotImplementedException>();
            exception = Record
                .ExceptionAsync( () =>  fail())
                .GetAwaiter()
                .GetResult();
            exception.Should().BeOfType<NotImplementedException>();
        }
```

**Review by vincentkam — APPROVED** [review]

**Comment by ksadoff** [comment]
Rebased, merged, and pushed.
